### ISSUE TITLE
Mean-Pooling Layer for Bert for varlen/nested/remove_input_padding tensor

### ISSUE THREAD

**Issue by michaelfeil** [NVIDIA/TensorRT-LLM#2554]
I am trying to implement a Bert-Style mean pooling using the Python API.

How would I implement this layer using tensorrt_llm abstractions?
https://github.com/NVIDIA/TensorRT-LLM/blob/340a1b62fc7a4124d6efef10df695b263734ab2f/tensorrt_llm/models/bert/model.py#L311

```python
from tensorrt_llm.functional import (
    ACT2FN,
    Tensor,
    concat,
    constant,
    cumsum,
    expand,
    index_select,
    select,
    shape,
    slice,
    unsqueeze,
    mean,
)
from tensorrt_llm.functional import sum as trt_sum

class MeanPooling(Module):
    """Custom Function, performs average over dimension 1 of a tensor"""

    def __init__(self):
        super().__init__()        

    def forward(self, hidden_states: "Tensor", input_lengths: "Tensor", remove_input_padding: bool, attention_mask: Optional["Tensor"], max_input_length: Optional[int]):
        """
        if not remove_input_padding:
            hidden_states: [batch_size, num_tokens, hidden_size]
            input_lengths: [batch_size]
            attention_mask: [batch_size, num_tokens, hidden_size] for 1=participate, 0=not participate
        elif remove_input_padding:
            hidden_states: [num_tokens, hidden_size]
            input_lengths: [batch_size]
            attention_mask: None
        """
        if not remove_input_padding:
            # if not remove_input_padding, the attention_mask is set and [batch_size, num_tokens, hidden_size] for 1=participate, 0=not participate
            # dimension is [batch_size, num_tokens, hidden_size]
            sum_hidden = trt_sum(hidden_states * attention_mask, dim=1) 
            # sum_hidden [batch_size, hidden_size]            
            mean_hidden = sum_hidden / unsqueeze(input_lengths, 1)
            return mean_hidden
        else:
            # state is [num_tokens, hidden_size]
            # e.g. input_lengths = [8, 5, 6]
            # hidden_states = [8+5+6=19, hidden_size]
            # mean_embeddings = [len(input_lengths), hidden_size]
            # mean_embeddings = [mean([hidden_states[0:8], dim=1), mean([hidden_states[8:13], dim=1), mean([hidden_states[13:19], dim=1)]
            # cumsum_indices = [0, 8, 13, 19]
            cumsum_indices = cumsum(
                concat(
                    [
                        0,
                        slice(
                            input_lengths,
                            starts=[0],
                            sizes=(
                                shape(input_lengths)
                                
                            ),
                        ),
                    ]
                ),
                0,
            )
            final_hidden_states = []
            for i in range(input_lengths.size(0)): # Does not work with the for-loop PSEUDOCODE
                # get the hidden states of the sequence
                sequence_hidden_states = slice(hidden_states, starts=[cumsum_indices[i]], sizes=[input_lengths[i]])
                # calculate the mean of the hidden states of the sequence
                mean_hidden_states = mean(sequence_hidden_states, dim=0, keepdim=True)
                final_hidden_states.append(mean_hidden_states)
            hidden_states = concat(final_hidden_states, 0)
            return hidden_states
```

**Comment by michaelfeil** [comment]
```python
class MeanPooling(Module):
    """Custom Function, performs average over the embed dim"""

    def __init__(self, hidden_size: int):
        super().__init__()        
        self.hidden_size = hidden_size

    def forward(self, hidden_states: "Tensor", input_lengths: "Tensor", remove_input_padding: bool, attention_mask: Optional["Tensor"], max_input_length: Optional[int]):
        """
        if not remove_input_padding:
            hidden_states: [batch_size, num_tokens, hidden_size]
            input_lengths: [batch_size]
            attention_mask: [batch_size, num_tokens, hidden_size] for 1=participate, 0=not participate
        elif remove_input_padding:
            hidden_states: [num_tokens, hidden_size]
            input_lengths: [batch_size]
            attention_mask: None
        """
        if not remove_input_padding:
            # if not remove_input_padding, the attention_mask is set and [batch_size, num_tokens, hidden_size] for 1=participate, 0=not participate
            # dimension is [batch_size, num_tokens, hidden_size]
            sum_hidden = trt_sum(hidden_states * attention_mask, dim=1) 
            # sum_hidden [batch_size, hidden_size]            
            mean_hidden = sum_hidden / trt_sum(attention_mask, 1)
            return mean_hidden
        else:
            # state is [num_tokens, hidden_size]
            # e.g. input_lengths = [8, 5, 6]
            # hidden_states = [8+5+6=19, hidden_size]
            # mean_embeddings = [len(input_lengths), hidden_size]
            # mean_embeddings = [mean([hidden_states[0:8], dim=1), mean([hidden_states[8:13], dim=1), mean([hidden_states[13:19], dim=1)]
            # cumsum_indices = [0, 8, 13, 19]
            offsets = cumsum(input_lengths, 0)

            cumsum_states_with_zero = concat(
                [
                    constant(np.zeros((1, self.hidden_size), dtype=np.float32)),
                    cumsum(cast(hidden_states, dtype="float32"), 0)
                ],
                dim=0 
            ) # size = [num_tokens+1, hidden_size]
            
            start_indices = offsets - input_lengths  # [batch_size, 1]
            end_indices = offsets  # [batch_size, 1]
            
            # go from a [num_tokens+1, hidden_size] tensor to a 2x [batch_size, hidden_size] tensor with the cumsum
            sum_start = index_select(cumsum_states_with_zero, index=start_indices, dim=1)
            sum_end = index_select(cumsum_states_with_zero, index=end_indices, dim=1)
             
            sum_hidden = (sum_end - sum_start) # [batch_size, hidden_size]
            
            mean_hidden = sum_hidden / cast(unsqueeze(input_lengths, 1), dtype="float32")
            
            return sum_hidden
```

```
rror Code: 9: Skipping tactic 0x0000000000000000 due to exception Assertion t.region->getDataType() == DataType::kINT32 failed. 
[12/11/2024-22:04:17] [TRT] [E] IBuilder::buildSerializedNetwork: Error Code 10: Internal Error (Could not find any implementation for node {ForeignNode[BertModel/layers/5/post_layernorm/value_L145/get_tensor_L136/get_constant_tensor_L127/_create_constant_tensor_L111/constant_L1150/CONSTANT_0 + BertModel/layers/5/post_layernorm/layer_norm_L5214/broadcast_helper_L2800/expand_dims_like_L1905/expand_dims_L1737/view_L1645/SHUFFLE_0...BertModel/norm_L619/__truediv___L357/elementwise_binary_L2855/ELEMENTWISE_DIV_0]}.)
```

**Comment by michaelfeil** [comment]
Mhh, padding mode implementation would likley be ~50-100h + worse performance over e.g. candle runtime. We have moved off trt for these models.
